### DISCUSSION TITLE
Configuration of cloud server and using data in QGIS

### DISCUSSION THREAD

**Discussion by jakubstaszel** [copcio/copcio.github.io#61] [Q&A]
Hello, I am trying to use COPC data in QGIS 3.28.7 from Azure Storage blob, but every time there is an error: 
Invalid Data Source: https://magisterkacog.blob.core.windows.net/aois/autzen-classified.copc.laz is not a valid or recognized data source, error: "Unable to recognize https://magisterkacog.blob.core.windows.net/aois/autzen-classified.copc.laz as a LAZ file: "The server of submitted URL doesn't support range queries""

The only thing that I did is that I moved your file from https://s3.amazonaws.com/hobu-lidar/autzen-classified.copc.laz to https://magisterkacog.blob.core.windows.net/aois/autzen-classified.copc.laz. Additionally, the data from S3 is accessible from QGIS, no error.

I configured CORS and the data is normally visible in COPC Viewer, the problem is only with QGIS. Can you tell me what additional configuration you did for S3?

**Comment by connormanning** [comment]
Your data works fine for me: [example](https://viewer.copc.io?q=https://magisterkacog.blob.core.windows.net/aois/autzen-classified.copc.laz).

So you'll have to ask the QGIS folks why QGIS thinks range requests are not supported when actually they are working.  There could be a number of reasons.  Maybe there is some preliminary check (like an OPTIONS/HEAD request etc.) that is hinting that Range requests might not be supported so QGIS is not actually continuing on to make any of the GET requests.

As far as I can tell, your endpoint seems configured properly - the key pieces are a) proper CORS and b) enabling Range requests (this was recently added as a [note](https://copc.io/#server-implementation-notes) in the spec).

Maybe @wonder-sk can chime in.

**Comment by uclaros** [comment]
QGIS relies on the server's reply Accept-Ranges header in order to load copc files.
It does a preliminary HEAD request with a Range: bytes=0-0 header and checks for the Accept-Ranges header in the reply.
It seems your server does support range requests but does not advertise so in the headers.

**Comment by wonder-sk** [comment]
As @uclaros says - QGIS currently requires that server sends `Accept-Ranges` header in the reply.

Looking at the RFC on range requests, this requirement is maybe a bit strict on QGIS side, as the RFC permits clients to do range requests even if this header is not present. In any case, it is a good idea to have `Accept-Ranges: bytes` header sent by the server.
https://datatracker.ietf.org/doc/html/rfc7233#section-2.3

**Comment by jakubstaszel** [comment]
Thank you for the advice! I will start the communication with QGIS community.
